server/src/main/java/org/jetbrains/bsp/bazel/server/sync/dependencytree/DependencyTree.kt

```diff
@@ -0,0 +1,96 @@
+package org.jetbrains.bsp.bazel.server.sync.dependencytree
+
+import org.jetbrains.bsp.bazel.info.BspTargetInfo.Dependency
+import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo
+
+
+class DependencyTree(
+        private val rootTargets: Set<String> = emptySet(),
+        private val idToTargetInfo: Map<String, TargetInfo> = emptyMap(),
+) {
+    private val idToDirectDependenciesIds: Map<String, Set<String>>
+    private val idToLazyTransitiveDependencies: Map<String, Lazy<Set<TargetInfo>>>
+
+    init {
+        idToDirectDependenciesIds = idToTargetInfo.entries.associate { (id, target) ->
+            Pair(
+                    id,
+                    getDependencies(target).filter(::exclude3rdPartyTargets).toSet()
+            )
+        }
+        idToLazyTransitiveDependencies = createIdToLazyTransitiveDependenciesMap(idToTargetInfo)
+    }
+
+    private fun exclude3rdPartyTargets(targetId: String) = targetId.startsWith("//")
+
+    private fun createIdToLazyTransitiveDependenciesMap(
+            idToTargetInfo: Map<String, TargetInfo>
+    ): Map<String, Lazy<Set<TargetInfo>>> {
+        return idToTargetInfo.mapValues { (_, targetInfo) ->
+            calculateLazyTransitiveDependenciesForTarget(targetInfo)
+        }
+    }
+
+    private fun calculateLazyTransitiveDependenciesForTarget(
+            targetInfo: TargetInfo
+    ): Lazy<Set<TargetInfo>> {
+        return lazy { calculateTransitiveDependenciesForTarget(targetInfo) }
+    }
+
+    private fun calculateTransitiveDependenciesForTarget(targetInfo: TargetInfo): Set<TargetInfo> {
+        val dependencies = getDependencies(targetInfo)
+        val strictlyTransitiveDependencies = calculateStrictlyTransitiveDependencies(dependencies)
+        val directDependencies = idsToTargetInfo(dependencies)
+        return strictlyTransitiveDependencies + directDependencies
+    }
+
+    private fun calculateStrictlyTransitiveDependencies(dependencies: Set<String>): Set<TargetInfo> {
+        return dependencies.flatMap { key ->
+            idToLazyTransitiveDependencies[key]?.value ?: listOf()
+        }.toSet()
+    }
+
+    private fun idsToTargetInfo(dependencies: Set<String>): Set<TargetInfo> {
+        return dependencies.mapNotNull(idToTargetInfo::get).toSet()
+    }
+
+    private fun directDependenciesIds(targetIds: Set<String>) =
+            targetIds.flatMap { targetId ->
+                idToDirectDependenciesIds.getOrDefault(
+                        targetId,
+                        setOf()
+                )
+            }.toSet()
+
+    fun allTargetsAtDepth(depth: Int, targets: Set<String>): Set<TargetInfo> =
+            idsToTargetInfo(targets) + when {
+                depth < 0 -> targets.flatMap { key ->
+                    idToLazyTransitiveDependencies[key]?.value ?: setOf()
+                }.filter { exclude3rdPartyTargets(it.id) }.toSet()
```

Comment:
Handling 3rd party like that feels unclear to me, because this is used in 2 places that don't feel related unless you well read into the code. Can we just filter deps at the end according to some predicate? Maybe dependency tree shouldn't bother with filtering if it is not important for performance.
I suggest 2 api variants (usage in BazelProjectMapper)
tree.targetsByDepth(rootTargets, depth) { isWorkspaceTarget(it) }
or
tree.targetsByDepth(rootTargets, depth).filter { isWorkspaceTarget(it) }
Do you think it can be implemented in optimal way?